contracts/DomainRegistry.sol

```diff
@@ -11,55 +16,99 @@ contract DomainRegistry {
         _;
     }
 
-    modifier domainDoesNotExist(string memory _topLevelDomain) {
-        require(!domains[_topLevelDomain].isRegistered, "Domain exists");
+ modifier domainExistence(string memory domainName, bool shouldExist) {
+    bool exists = domains[domainName.stripProtocol()].owner != address(0);
+    if (shouldExist) {
+        require(exists, "Domain doesn't exist");
+    } else {
+        require(!exists, "Domain exists");
+    }
+    _;
+}
+
+    modifier isValidDomain(string memory _domainName) {
+        _domainName = _domainName.stripProtocol();
+        require(bytes(_domainName).length > 0, "Domain is empty");
+        bytes memory domainBytes = bytes(_domainName);
+        require(domainBytes[domainBytes.length - 1] != bytes1('.'), "Domain ends with a dot");
         _;
     }
 
-    modifier domainExists(string memory _topLevelDomain) {
-        require(domains[_topLevelDomain].isRegistered, "Domain doesn't exist");
+    modifier domainOwnedBySender(string memory domainName) {
+        domainName = domainName.stripProtocol();
+        require(domains[domainName].owner == msg.sender, "Not the domain owner");
         _;
     }
 
-    modifier domainOwnedBySender(string memory _topLevelDomain) {
-        require(domains[_topLevelDomain].owner == msg.sender, "Not the domain owner");
+modifier parentDomainExists(string memory domainName) {
+    if (isTopLevelDomain(domainName)) {
         _;
+        return;
+    }
+    
+    string memory parentDomain = domainName.extractParentDomain();
+    if (bytes(parentDomain).length > 1) {  
+        require(domains[parentDomain].isRegistered, "Parent domain doesn't exist");
     }
+    _;
+}
 
-    modifier isTopLevelDomain(string memory _topLevelDomain) {
-        require(bytes(_topLevelDomain).length > 0, "Domain is empty");
-        for (uint i = 0; i < bytes(_topLevelDomain).length; i++) {
-            require(bytes(_topLevelDomain)[i] != bytes(".")[0], "Multilevel domains are not allowed");
+function isTopLevelDomain(string memory domainName) internal pure returns (bool) {
+    bytes memory domainBytes = bytes(domainName);
+    uint dotCount = 0;
+    for (uint i = 0; i < domainBytes.length; i++) {
+        if (domainBytes[i] == bytes1('.')) {
+            dotCount++;
         }
-        _;
     }
+    return dotCount == 0;  
+}
 
-    function registerDomain(string memory _topLevelDomain) public hasRequiredDeposit() domainDoesNotExist(_topLevelDomain) isTopLevelDomain(_topLevelDomain) payable {
-        domains[_topLevelDomain] = Domain({
-            deposit: msg.value,
-            isRegistered: true,
-            owner: msg.sender
-        });
+function registerDomain(string memory domainName) 
+    public 
+    hasRequiredDeposit() 
+    domainExistence(domainName, false) 
+    isValidDomain(domainName)
+    parentDomainExists(domainName)
+    payable 
+{
+    domainName = domainName.stripProtocol();
+    require(bytes(domainName).length > 0, "Domain cannot be empty");
```

Comment:
You repeat the same check here and in the modifier `isValidDomain(domainName)` - 1 point

Reply:
fxd :(